technical-docs/K8SSANDRA-OPERATOR-DESIGN.md

```diff
@@ -0,0 +1,410 @@
+# K8ssandra Operator Design
+
+Author: @jsanda
+
+## Overview
+
+GH issue: https://github.com/k8ssandra/k8ssandra/issues/683
+
+This is a design document for an operator for k8ssandra. For background see 
+https://github.com/k8ssandra/k8ssandra/issues/485.
+
+Much of the behavior that is implemented in the Helm charts will be reimplemented in various controllers. The role of the Helm charts changes to focus primarily on installation and configuration of operators.
+
+While we will continue to use and support Helm, it will not be required for installing k8ssandra from a purely technical perspective. Users can install manifests with kubectl apply for example. We can also provide support for kustomize.
+
+## CustomResourceDefinitions
+
+We will utilize a number of CRDs, some of which are developed and maintained by other projects.
+
+There will be a separate controller for each CRD.
+
+The goal of this document is not to be entirely prescriptive for each CRD. Rather, the intent is to give a general idea of how things will look and fit together. 
+
+### K8ssandraCluster
+
+This is the primary object which users will manage. The controller will be responsible for creating other constituent objects.
+
+It will look something like this:
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: K8ssandraCluster
+metadata:
+  name: k8ssandra
+spec:
+  cassandra:
+    clusterName: k8ssandra
+    auth:
+      enabled: true
+    datacenters:
+    - name: dc1
+      size: 3
+  reaper:
```

Comment:
One more discussion point.. do we still want to maintain that all these are separate products? Or just separate features? Could I just configure my repairs here? My backups here? I know we've gone back and forth with these, but does the word "reaper" give anything to users? Or medusa? Or stargate?

Reply:
Or just maintenance-tools ?